lib/nodes/proxyShapePlugin.cpp

```diff
@@ -0,0 +1,138 @@
+int _registrationCount = 0;
+
+// Name of the plugin registering the proxy shape base class.
+MString _registrantPluginName;
+
+bool _useVP2RenderDelegate = false;
+
+TF_DEFINE_ENV_SETTING(VP2_RENDER_DELEGATE_PROXY, false,
+    "Switch proxy shape rendering to VP2 render delegate.");
+}
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+/* static */
+MStatus
+MayaUsdProxyShapePlugin::initialize(MFnPlugin& plugin)
+{
+    // If we're already registered, do nothing.
+    if (_registrationCount++ > 0) {
+        return MS::kSuccess;
+    }
+
+    _registrantPluginName = plugin.name();
+
+    _useVP2RenderDelegate = TfGetEnvSetting(VP2_RENDER_DELEGATE_PROXY);
+
+    MStatus status;
+
+    // Proxy shape initialization.
+    status = plugin.registerData(
+        MayaUsdStageData::typeName,
+        MayaUsdStageData::mayaTypeId,
+        MayaUsdStageData::creator);
+    CHECK_MSTATUS(status);
+
+    status = plugin.registerShape(
+        MayaUsdProxyShapeBase::typeName,
+        MayaUsdProxyShapeBase::typeId,
+        MayaUsdProxyShapeBase::creator,
+        MayaUsdProxyShapeBase::initialize,
+        nullptr,
+        getProxyShapeClassification());
+    CHECK_MSTATUS(status);
+
+        status = MHWRender::MDrawRegistry::registerSubSceneOverrideCreator(
+            ProxyRenderDelegate::drawDbClassification,
+            _RegistrantId,
+            ProxyRenderDelegate::Creator);
+        CHECK_MSTATUS(status);
+
+    status = HdVP2ShaderFragments::registerFragments();
+    CHECK_MSTATUS(status);
+
+    return status;
+}
+
+/* static */
+MStatus
+MayaUsdProxyShapePlugin::finalize(MFnPlugin& plugin)
+{
+    // If more than one plugin still has us registered, do nothing.
+    if (_registrationCount-- > 1) {
+        return MS::kSuccess;
+    }
+
+    // Maya requires deregistration to be done by the same plugin that
+    // performed the registration.  If this isn't possible, warn and don't
+    // deregister.
+    if (plugin.name() != _registrantPluginName) {
+        MGlobal::displayWarning(
+            "USD proxy shape base cannot be deregistered, registering plugin "
+            + _registrantPluginName + " is unloaded.");
+        return MS::kSuccess;
+    }
+
+    MStatus status = HdVP2ShaderFragments::deregisterFragments();
+    CHECK_MSTATUS(status);
+    
+        status = MHWRender::MDrawRegistry::deregisterSubSceneOverrideCreator(
+            ProxyRenderDelegate::drawDbClassification,
+            _RegistrantId);
+        CHECK_MSTATUS(status);
+
+    status = plugin.deregisterNode(MayaUsdProxyShapeBase::typeId);
+    CHECK_MSTATUS(status);
+    
+    status = plugin.deregisterData(MayaUsdStageData::mayaTypeId);
+    CHECK_MSTATUS(status);
+
+    return status;
+}
+
+const MString* MayaUsdProxyShapePlugin::getProxyShapeClassification()
+{
+    return _useVP2RenderDelegate ? &ProxyRenderDelegate::drawDbClassification : 
+        nullptr;
+}
+
+bool MayaUsdProxyShapePlugin::useVP2_NativeUSD_Rendering()
+{
+    return _useVP2RenderDelegate;
+}
```

Comment:
Inlining this function, or similarly exposing the underlying data member by making it public, breaks encapsulation and exposes implementation details.  In this case, it doesn't really matter that much, but it's still good practice.  In general inlining should be reserved for time-critical code (and this definitely isn't).

Reply:
I have to say I disagree. Ok, so the code for the method will probably end up being something boring like:
```asm
        movzx   eax, BYTE PTR [rdi + someOffset]
        ret
```
Not all that bad, in of itself. However. This is now a DSO exported method, which means the compiler cannot see if the method has any side effects (e.g. does it modify any registers, does it do something to a file on disk, print an error message, etc). 

This has implications for the place where we call that function.... 

1. The compiler cannot see behind the DSO boundary. 
2. The x64 convention requires the caller to save/restore the volatile registers, which is anywhere up to 19 or so registers on AVX+ machines (and approx 43 registers for AVX512). 
3. So the ultimate runtime performance of this method is now dependent on the complexity of the code surrounding the call-site (i.e. how many registers are currently inflight)

In the absolute worst case scenario (on AVX512), this simple getter could end up forcing the compiler to emit anywhere up to ~90 CPU ops each time you call this function _(Admittedly it would be extremely hard to achieve that, but it's not impossible!)._ 

If however you inlined that method, the runtime cost will be zero ops in the best case, and one op in the worst case (x64 can fuse loads into most instructions, e.g. add/sub/etc). 

Add on to that the fact the compiler cannot see behind the DSO barrier, which will pretty much kill any auto-vectorisation or loop unrolling optimisations that may have previously been possible. 

Whilst encapsulation is a nice tool for maintaining ABI compatibility, I'm not entirely convinced that's a required aim for what is effectively a plugin?

Reply:
That's a fine performance analysis, but again this function is called once, at initialization, so I don't believe it's time-sensitive at all.  In terms of ABI, we're building a mayaUsd library to support multiple plugins, so although it's not like we're building an interface used by millions of people, it's still reasonable to try to get it right and provide encapsulation unless performance requires it.